File: ANativeProject/codec_native/src/main/java/com/lmy/hwvcnative/processor/AudioProcessor.kt
package com.lmy.hwvcnative.processor

import android.view.Surface
import com.lmy.hwvcnative.CPPObject
import com.lmy.hwvcnative.FilterSupport
import com.lmy.hwvcnative.filter.Filter

class AudioProcessor : CPPObject() {
    private var filter: Filter? = null

    init {
        handler = create()
    }

    fun setSource(path: String) {
        if (0L == handler) return
        setSource(handler, path);
    }

    fun prepare() {
        if (0L == handler) return
        prepare(handler)
    }

    fun start() {
        if (0L == handler) return
        start(handler)
    }

    fun pause() {
        if (0L == handler) return
        pause(handler)
    }

    fun stop() {
        if (0L == handler) return
        stop(handler)
    }

    fun seek(us: Long) {
        if (0L == handler) return
        seek(handler, us)
    }

    fun release() {
        release(handler)
        handler = 0
    }

    private external fun create(): Long
    private external fun setSource(handler: Long, path: String)
    private external fun prepare(handler: Long)
    private external fun start(handler: Long)
    private external fun pause(handler: Long)
    private external fun stop(handler: Long)
    private external fun seek(handler: Long, us: Long)
    private external fun release(handler: Long)
}

File: ANativeProject/hwvc/src/render/processor/AudioProcessor.cpp
/*
* Copyright (c) 2018-present, lmyooyo@gmail.com.
*
* This source code is licensed under the GPL license found in the
* LICENSE file in the root directory of this source tree.
*/

#include "../include/AudioProcessor.h"
#include "../include/HwAudioInput.h"
#include "ObjectBox.h"

AudioProcessor::AudioProcessor() : Object() {
    pipeline = new UnitPipeline("AudioProcessor");
    pipeline->registerAnUnit(new HwAudioInput());
}

AudioProcessor::~AudioProcessor() {
    if (pipeline) {
        pipeline->release();
        wait(20000);
        delete pipeline;
        pipeline = nullptr;
    }
}

void AudioProcessor::setSource(char *path) {
    if (pipeline) {
        Message *msg = new Message(EVENT_AUDIO_SET_SOURCE, nullptr);
        msg->obj = new ObjectBox(path);
        pipeline->postEvent(msg);
    }
}

void AudioProcessor::prepare() {
    if (pipeline) {
        Message *msg = new Message(EVENT_COMMON_PREPARE, nullptr);
        msg->obj = new ObjectBox(nullptr);
        pipeline->postEvent(msg);
    }
}

void AudioProcessor::start() {
    if (pipeline) {
        Message *msg = new Message(EVENT_AUDIO_START, nullptr);
        pipeline->postEvent(msg);
    }
}

void AudioProcessor::pause() {
    if (pipeline) {
        Message *msg = new Message(EVENT_AUDIO_PAUSE, nullptr);
        pipeline->postEvent(msg);
    }
}

void AudioProcessor::stop() {
    if (pipeline) {
        Message *msg = new Message(EVENT_AUDIO_STOP, nullptr);
        pipeline->postEvent(msg);
    }
}

void AudioProcessor::seek(int64_t us) {
    if (pipeline) {
        pipeline->removeAllMessage(EVENT_AUDIO_SEEK);
        Message *msg = new Message(EVENT_AUDIO_SEEK, nullptr);
        msg->arg2 = us;
        pipeline->postEvent(msg);
    }
}